### PR TITLE
Fix compile error for CUDA in TRMM and TRTRI perf tests

### LINKED ISSUE #727: trmm perf test nightly build error: clang/8+cuda tpl build
The following nightly test failed on kokkos-dev-2 after merge of PR #720, @e10harvey can you take a look? From the compiler message looks like either a routine is missing a `KOKKOS_INLINE_FUNCTION` marking, or a host function is being incorrectly called from a kernel in a `Kokkos::parallel_for`

clang-8.0-Cuda_OpenMP-release (build failed)
https://jenkins-son.sandia.gov/job/KokkosKernels_KokkosDev2_SPOTCHECK_TPLS/83/console

Other impacted builds:
https://jenkins-son.sandia.gov/view/KokkosKernels/job/KokkosKernels_KokkosDev2_CLANG8_CUDA10_cpp17/
https://jenkins-son.sandia.gov/view/KokkosKernels/job/KokkosKernels_KokkosDev2_CLANG8_CUDA10_cpp17/
https://jenkins-son.sandia.gov/view/KokkosKernels/job/KokkosKernels_KokkosDev_CLANG7_CUDA9/84/console

**First compile error**
```bash
In file included from /home/jenkins/slave/workspace/KokkosKernels_KokkosDev2_SPOTCHECK_TPLS/kokkos-kernels/perf_test/blas/blas3/KokkosBlas3_perf_test.cpp:45:
/home/jenkins/slave/workspace/KokkosKernels_KokkosDev2_SPOTCHECK_TPLS/kokkos-kernels/perf_test/blas/blas3/KokkosBlas3_trmm_perf_test.hpp:303:17: error: reference to __host__ function 'trmm<Kokkos::View<double **, Kokkos::LayoutStride, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0> >, Kokkos::View<double **, Kokkos::LayoutStride, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0> > >' in __host__ __device__ function
 KokkosBlas::trmm(&trmm_args_.side, &trmm_args_.uplo, &trmm_args_.trans,
```

**Reproducer instructions:**
```bash
   #   Load modules:
         source /projects/sems/modulefiles/utils/sems-modules-init.sh ; module use /home/projects/x86-64/modulefiles/local
         module purge
         module load sems-env sems-cmake/3.12.2 kokkos-env clang/8.0 cuda/10.0
 
       $KOKKOSKERNELS_PATH/cm_generate_makefile.bash --with-devices=Cuda,OpenMP --arch=SNB,Volta70 --compiler=/home/projects/x86-64/clang/8.0/bin/clang++ --cxxflags="-O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-pass-failed " --cxxstandard="11" --ldflags="" --with-cuda=/home/projects/x86-64/cuda/10.0 --kokkos-path=$KOKKOS_PATH --kokkoskernels-path=$KOKKOSKERNELS_PATH --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --with-tpls=blas    --with-options= --with-cuda-options= --no-examples 
```

Edit: Added links to other affected builds

### PR DESCRIPTION
**Pull request by e10harvey**
- Also fixes:
  - `-r` command line options
  - `__do_setup` for CUDA.

Fixes #727.

Running spot checks now.

### CODE REVIEW

**Comment by e10harvey** [comment]
Hang on. We passed spot-check-tpls without these changes. This may be related to a change in kokkos. I need to investigate #727 further.

**Comment by e10harvey** [comment]
> Hang on. We passed spot-check-tpls without these changes. This may be related to a change in kokkos. I need to investigate #727 further.

@ndellingwood: This is ready to review.

I tried spot-check-tpls with e10harvey/issue-711 again and got the build error that #727 reports. This should've been caught while spot checking #720. #654 should be resolved soon.

The spot check passed the build but unit tests are failing with unaligned memory accesses. Those test failures are not related to the TRMM and TRTRI perf tests.

```
$ tail -4 Cuda_OpenMP-release/*.test.log
The following tests FAILED:
          1 - blas_cuda (Child aborted)
          3 - sparse_cuda (Child aborted)
Errors while running CTest
```

```
$ tail -4 Cuda_Serial-release/*.test.log
The following tests FAILED:
          1 - blas_cuda (Child aborted)
          3 - sparse_cuda (Child aborted)
Errors while running CTest
```

**Comment by ndellingwood** [comment]
> Those test failures are not related to the TRMM and TRTRI perf tests.

@e10harvey thanks for the update, can you post a few more items:

* What SHA of kokkos are you building against? Alignment issues should hopefully be resolved since tagged release branch 3.1.1
* Which of the blas and sparse tests failed due to alignment? Luc noticed a gemv failure in tests that was resolved by using an updated pull from kokkos and previous failures Brian documented were resolved with kokkos/3.1.1, hopefully nothing new has popped up

File: perf_test/blas/blas/KokkosBlas_trtri_perf_test.hpp [hunk 1]
```diff
@@ -320,19 +329,23 @@ trtri_args_t __do_setup(options_t options, matrix_dims_t dim) {
   uint64_t seed = Kokkos::Impl::clock_tic();
   Kokkos::Random_XorShift64_Pool<execution_space> rand_pool(seed);
   decltype(dim.a.m) min_dim = dim.a.m < dim.a.n ? dim.a.m : dim.a.n;
+  typename vta::HostMirror host_A;
   STATUS;
 
   trtri_args.uplo = options.blas_args.trtri.trtri_args.c_str()[0];
   trtri_args.diag = options.blas_args.trtri.trtri_args.c_str()[1];
   trtri_args.A    = vta("trtri_args.A", options.n, dim.a.m, dim.a.n);
+  host_A = Kokkos::create_mirror_view(trtri_args.A);
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_trtri_perf_test.hpp:338]: @e10harvey you should be able to create `host_A` via `auto` type deduction, I don't think the declaration at line 332 above is necessary unless you prefer this pattern

File: perf_test/blas/blas3/KokkosBlas3_trmm_perf_test.hpp [hunk 2]
```diff
@@ -492,14 +503,17 @@ trmm_args_t __do_setup(options_t options, matrix_dims_t dim) {
   trmm_args.diag  = options.blas_args.trmm.trmm_args.c_str()[3];
   trmm_args.A     = vta("trmm_args.A", options.n, dim.a.m, dim.a.n);
   trmm_args.B     = vtb("trmm_args.B", options.n, dim.b.m, dim.b.n);
+  host_A = Kokkos::create_mirror_view(trmm_args.A);
```
Comment (ndellingwood) [inline perf_test/blas/blas3/KokkosBlas3_trmm_perf_test.hpp:506]: @e10harvey you should be able to create `host_A` via `auto` type deduction, I don't think the declaration at line 497 above is necessary unless you prefer this pattern (similar other comment)

**Comment by e10harvey** [comment]
@ndellingwood: Thanks for your feedback!

>     * What SHA of kokkos are you building against? Alignment issues should hopefully be resolved since tagged release branch 3.1.1

This was using `kokkos:3.0.00`.

>     * Which of the blas and sparse tests failed due to alignment? Luc noticed a gemv failure in tests that was resolved by using an updated pull from kokkos and previous failures Brian documented were resolved with kokkos/3.1.1, hopefully nothing new has popped up

The following tests failed using `kokkos:3.0.00` and `e10harvey:issue-727`:
- For both Serial and OpenMP:
  - From KokkosKernels_blas_cuda:
    - cuda.gemv_complex_double
  - From KokkosKernels_sparse_cuda:
    - cuda.sparse_gauss_seidel_asymmetric_rank1_kokkos_complex_double_int_int_TestExecSpace

Spot check results from last night using the tip of `kokkos:develop` and the tip of `e10harvey:issue-727`:
```
KokkosKernels Repository Status:  461e175724b555317069328d47e0dd192c1b9bce perf_test/blas/blas: Fix -r option

Kokkos Repository Status:  e526f7e26358ff458edb93934e1deea7d225fcd0 Merge pull request #3058 from dalg24/duplicated_code_in_block_size_deduction


Going to test compilers:  clang/8.0
Testing compiler clang/8.0
  Starting job clang-8.0-Cuda_OpenMP-release
kokkos devices: Cuda,OpenMP
kokkos arch: SNB,Volta70
kokkos options:
kokkos cuda options:
kokkos cxxflags: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-pass-failed
extra_args:
kokkoskernels scalars: 'double,complex_double'
kokkoskernels ordinals: int
kokkoskernels offsets: int,size_t
kokkoskernels layouts: LayoutLeft
  PASSED clang-8.0-Cuda_OpenMP-release
  Starting job clang-8.0-Pthread_Serial-release
kokkos devices: Pthread,Serial
kokkos arch: SNB,Volta70
kokkos options:
kokkos cuda options:
kokkos cxxflags: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-pass-failed
extra_args:
kokkoskernels scalars: 'double,complex_double'
kokkoskernels ordinals: int
kokkoskernels offsets: int,size_t
kokkoskernels layouts: LayoutLeft
  PASSED clang-8.0-Pthread_Serial-release
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=618 run_time=130
clang-8.0-Pthread_Serial-release build_time=202 run_time=110
```

**Pull request merged**